`solana-unified-scheduler-pool` fails to build in nightly rust

### Problem

This library now fails to build when using recent versions of nightly rust due to a change in the `extract_if` API. 

https://github.com/rust-lang/rust/commit/ab6ad1ae0616fbfffa73efe1cb702150c4c7f7d1

https://github.com/rust-lang/rust/blame/c3dd4eefd64da263865f28d4c39d0ef09433d5bf/library/alloc/src/vec/mod.rs#L3706

In nightly rust the built-in `extract_if` supersedes the polyfilled trait.

https://github.com/anza-xyz/agave/blob/4b9b0b7786d9aa5693984249be7b9472c00cc998/unified-scheduler-pool/src/lib.rs#L383-L393

### Proposed Solution

I'll create a pull request to directly call the polyfill trait method to remove ambiguity and retain support for nightly and stable. 

### Error

```bash
error[E0282]: type annotations needed for `(_, _)`
   --> /Users/user/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/solana-unified-scheduler-pool-2.1.15/src/lib.rs:214:69
    |
214 | ...cheduler_inners.extract_if(|(_inner, pooled_at)| {
    |                                ^^^^^^^^^^^^^^^^^^^
215 | ...ce(*pooled_at) > max_pooling_duration
    |       ---------- type must be known at this point
    |
help: consider giving this closure parameter an explicit type, where the placeholders `_` are specified
    |
214 |                     idle_inners.extend(scheduler_inners.extract_if(|(_inner, pooled_at): (_, _)| {
    |                                                                                        ++++++++

error[E0061]: this method takes 2 arguments but 1 argument was supplied
   --> /Users/user/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/solana-unified-scheduler-pool-2.1.15/src/lib.rs:214:57
    |
214 |                       idle_inners.extend(scheduler_inners.extract_if(|(_...
    |  _________________________________________________________^^^^^^^^^^-
215 | |                         now.duration_since(*pooled_at) > max_pooling_d...
216 | |                     }));
    | |______________________- argument #2 is missing
    |
note: method defined here
   --> /rustc/4d30011f6c616be074ba655a75e5d55441232bbb/library/alloc/src/vec/mod.rs:3706:12
help: provide the argument
    |
216 |                     idle_inners.extend(scheduler_inners.extract_if(|(_inner, pooled_at)| {
217 |                         now.duration_since(*pooled_at) > max_pooling_duration
218 ~                     }, /* filter */));
    |

error[E0061]: this method takes 2 arguments but 1 argument was supplied
   --> /Users/user/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/solana-unified-scheduler-pool-2.1.15/src/lib.rs:245:64
    |
245 |                       expired_listeners.extend(timeout_listeners.extract_if(
    |  ________________________________________________________________^^^^^^^^^^-
246 | |                         |(_callback, registered_at)| {
247 | |                             now.duration_since(*registered_at) > timeout_dura...
248 | |                         },
249 | |                     ));
    | |_____________________- argument #2 is missing
    |
note: method defined here
   --> /rustc/4d30011f6c616be074ba655a75e5d55441232bbb/library/alloc/src/vec/mod.rs:3706:12
help: provide the argument
    |
245 ~                     expired_listeners.extend(timeout_listeners.extract_if(|(_callback, registered_at)| {
246 +                             now.duration_since(*registered_at) > timeout_duration
247 ~                         }, /* filter */));
    |

Some errors have detailed explanations: E0061, E0282.
For more information about an error, try `rustc --explain E0061`.
error: could not compile `solana-unified-scheduler-pool` (lib) due to 3 previous errors
warning: build failed, waiting for other jobs to finish..
```

--- a/unified-scheduler-pool/src/lib.rs
+++ b/unified-scheduler-pool/src/lib.rs
@@ -12,6 +12,8 @@
 //! Refer to [`PooledScheduler`] doc comment for general overview of scheduler state transitions
 //! regarding to pooling and the actual use.
 
+use std::ops::DerefMut;
+
 #[cfg(feature = "dev-context-only-utils")]
 use qualifier_attr::qualifiers;
 use {
@@ -387,9 +389,8 @@ where
                     //
                     // Note that this critical section could block the latency-sensitive replay
                     // code-path via ::take_scheduler().
-                    #[allow(unstable_name_collisions)]
                     idle_inners.extend(MakeExtractIf::extract_if(
-                        &mut *scheduler_inners,
+                        scheduler_inners.deref_mut(),
                         |(_inner, pooled_at)| now.duration_since(*pooled_at) > max_pooling_duration,
                     ));
                     drop(scheduler_inners);
@@ -419,9 +420,8 @@ where
                     let Ok(mut timeout_listeners) = scheduler_pool.timeout_listeners.lock() else {
                         break;
                     };
-                    #[allow(unstable_name_collisions)]
                     expired_listeners.extend(MakeExtractIf::extract_if(
-                        &mut *timeout_listeners,
+                        timeout_listeners.deref_mut(),
                         |(_callback, registered_at)| {
                             now.duration_since(*registered_at) > timeout_duration
                         },

--- a/unified-scheduler-pool/src/lib.rs
+++ b/unified-scheduler-pool/src/lib.rs
@@ -388,9 +388,10 @@ where
                     // Note that this critical section could block the latency-sensitive replay
                     // code-path via ::take_scheduler().
                     #[allow(unstable_name_collisions)]
-                    idle_inners.extend(scheduler_inners.extract_if(|(_inner, pooled_at)| {
-                        now.duration_since(*pooled_at) > max_pooling_duration
-                    }));
+                    idle_inners.extend(MakeExtractIf::extract_if(
+                        &mut *scheduler_inners,
+                        |(_inner, pooled_at)| now.duration_since(*pooled_at) > max_pooling_duration,
+                    ));
                     drop(scheduler_inners);
 
                     let idle_inner_count = idle_inners.len();
@@ -419,7 +420,8 @@ where
                         break;
                     };
                     #[allow(unstable_name_collisions)]
-                    expired_listeners.extend(timeout_listeners.extract_if(
+                    expired_listeners.extend(MakeExtractIf::extract_if(
+                        &mut *timeout_listeners,
                         |(_callback, registered_at)| {
                             now.duration_since(*registered_at) > timeout_duration
                         },

--- a/unified-scheduler-pool/src/lib.rs
+++ b/unified-scheduler-pool/src/lib.rs
@@ -12,8 +12,6 @@
 //! Refer to [`PooledScheduler`] doc comment for general overview of scheduler state transitions
 //! regarding to pooling and the actual use.
 
-use std::ops::DerefMut;
-
 #[cfg(feature = "dev-context-only-utils")]
 use qualifier_attr::qualifiers;
 use {
@@ -53,6 +51,7 @@ use {
         fmt::Debug,
         marker::PhantomData,
         mem,
+        ops::DerefMut,
         sync::{
             atomic::{AtomicU64, AtomicUsize, Ordering::Relaxed},
             Arc, Mutex, MutexGuard, OnceLock, Weak,

